File: src/Android/Tiles/MyVaultTileService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Android.App;
using Android.Content;
using Android.OS;
using Android.Runtime;
using Android.Service.QuickSettings;
using Android.Views;
using Android.Widget;
using Java.Lang;

namespace Bit.Droid.Tile
{
    [Service(Permission = Android.Manifest.Permission.BindQuickSettingsTile, Label = "@string/MyVault",
        Icon = "@drawable/shield")]
    [IntentFilter(new string[] { ActionQsTile })]
    [Register("com.x8bit.bitwarden.MyVaultTileService")]
    public class MyVaultTileService : TileService
    {
        public override void OnTileAdded()
        {
            base.OnTileAdded();
        }

        public override void OnStartListening()
        {
            base.OnStartListening();
        }

        public override void OnStopListening()
        {
            base.OnStopListening();
        }

        public override void OnTileRemoved()
        {
            base.OnTileRemoved();
        }

        public override void OnClick()
        {
            base.OnClick();

            if (IsLocked)
            {
                UnlockAndRun(new Runnable(() =>
                {
                    LaunchMyVault();
                }));
            }
            else
            {
                LaunchMyVault();
            }
        }

        private void LaunchMyVault()
        {
            var intent = new Intent(this, typeof(MainActivity));
            intent.SetFlags(ActivityFlags.NewTask | ActivityFlags.SingleTop | ActivityFlags.ClearTop);
            intent.PutExtra("myVaultTile", true);
            StartActivityAndCollapse(intent);
        }
    }
}

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/tiles/MyVaultTileService.kt
package com.x8bit.bitwarden.ui.platform.feature.tiles

import android.annotation.SuppressLint
import android.os.Build
import android.service.quicksettings.TileService
import com.x8bit.bitwarden.ui.platform.manager.intent.IntentManager
import com.x8bit.bitwarden.ui.platform.manager.intent.IntentManagerImpl
import kotlinx.coroutines.Runnable

/**
 * A service for handling the My Vault quick settings tile.
 */
class MyVaultTileService : TileService() {
    private val tileIntentManager: IntentManager = IntentManagerImpl(this)

    override fun onTileAdded() {
        super.onTileAdded()
    }

    override fun onStartListening() {
        super.onStartListening()
    }

    override fun onStopListening() {
        super.onStopListening()
    }

    override fun onTileRemoved() {
        super.onTileRemoved()
    }

    override fun onClick() {
        super.onClick()

        if (isLocked) {
            unlockAndRun(Runnable { launchVault(); })
        } else {
            launchVault()
        }
    }

    @Suppress("DEPRECATION")
    @SuppressLint("StartActivityAndCollapseDeprecated")
    private fun launchVault() {
        val intent = tileIntentManager.createTileIntent("bitwarden://my_vault")

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            startActivityAndCollapse(tileIntentManager.createTilePendingIntent(intent))
        } else {
            startActivityAndCollapse(intent)
        }
    }
}
